csharp/src/Google.Protobuf/JsonTokenizer.cs

```diff
@@ -728,6 +731,59 @@ internal InvalidJsonException CreateException(string message)
                     return new InvalidJsonException(message);
                 }
             }
+
+            /// <summary>
+            /// Provide a cached reusable instance of stringbuilder per thread.
+            /// Copied from https://github.com/dotnet/runtime/blob/main/src/libraries/Common/src/System/Text/StringBuilderCache.cs
+            /// </summary>
+            private static class StringBuilderCache
+            {
+                private const int MaxCachedStringBuilderSize = 360;
+                private const int DefaultStringBuilderCapacity = 16; // == StringBuilder.DefaultCapacity
+
+                [ThreadStatic]
+                private static StringBuilder cachedInstance;
+
+                /// <summary>Get a StringBuilder for the specified capacity.</summary>
+                /// <remarks>If a StringBuilder of an appropriate size is cached, it will be returned and the cache emptied.</remarks>
+                public static StringBuilder Acquire(int capacity = DefaultStringBuilderCapacity)
+                {
+                    if (capacity <= MaxCachedStringBuilderSize)
+                    {
+                        StringBuilder sb = cachedInstance;
+                        if (sb != null)
+                        {
+                            // Avoid stringbuilder block fragmentation by getting a new StringBuilder
+                            // when the requested size is larger than the current capacity
+                            if (capacity <= sb.Capacity)
+                            {
+                                cachedInstance = null;
+                                sb.Clear();
+                                return sb;
+                            }
+                        }
+                    }
+
+                    return new StringBuilder(capacity);
+                }
+
+                /// <summary>Place the specified builder in the cache if it is not too big.</summary>
+                private static void Release(StringBuilder sb)
+                {
+                    if (sb.Capacity <= MaxCachedStringBuilderSize)
+                    {
+                        cachedInstance = sb;
```

Comment:
Should we keep the bigger of the two here? There's a chance the cached one has bigger capacity than the one being released and we might want to keep the bigger one to reduce the chances of line 758 evaluating to true?

Reply:
The idea is to not keep too big StringBuilder instances as they are static per thread and can increase total memory consumption of the app. At least this are considerations that people in .net runtime had. As written above the class - this is a copy of internal class used there.

Reply:
Yes, I'm not saying we should keep anything that's above `MaxCachedStringBuilderSize`. I'm saying we should do something like the following code, to make sure that we keep cache the biggest string builder possible of the ones we have already allocated, within the `MaxCachedStringBuilderSize` boundaries.

```
cachedInstance = cachedInstance?.Capacity >= sb.Capacity ? cachedInstance : sb;
```

Reply:
If you check the code in Acquire - it clears cachedInstance and set it to null in happy path.
So when releasing **cachedInstance** will be null and there is no need to check it's capacity.

Reply:
I'm refering to this scenario:

- Acquiere sb of 50 capacity
- Release sb of 50 capacity => sb of 50 is now cached
- Acquire sb of 100 capacity => new sb of 100 is created, sb of 50 continues to be cached
- Release sb of 100 capacity => sb of 100 is discarded, even though it's better to discard the 50 capacity and keep the 100 capacity
- Acquire sb of 70 capacity => new sb of 70 capacity is created, sb of 50 continues to be cached, we could have reused the sb of 100 capacity

It's probably easy to write a test that demontrates this, unless I'm missunderstanding something. This seems like a simple change that would maximize the benefits of caching.